### ISSUE TITLE
[2020 CVPR Oral] Momentum Contrast for Unsupervised Visual Representation Learning

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#51]
MoCo v1 논문

Method, experiments, ablations만 설명

## 1. Method
<img src='https://user-images.githubusercontent.com/49643709/152975089-79a16c8f-5570-48fa-b3a3-9cb4f9a04d3a.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/152980199-bd737e85-dbde-4ff3-8ca9-6f548fec28d0.png' width=80% height=80%>

How MoCo works
1. Mini-batch의 이미지가 주어졌을 때, augmentation을 진행하여 query, key images를 얻음
2. Query image는 query encoder에, key image는 key encoder에 넣어 query, key representation을 얻음
3. Key representation을 얻을 때 batch를 shuffle해줌
4. Query representation과 key representation을 dot product를 통해 query와 key의 similarity를 계산
5. Positive에 해당하는 대각 성분 값을 추출하여 positive logit을 구함
6. Dictionary에 있는 queue representation과 query representation의 similarity를 구해 negative logit을 구함
7. Positive logit과 negative logit을 concat하여 cross entropy를 통해 loss 구함
8. Loss를 통해 query encoder update, key encoder update
9. Key encoder는 back-prop을 통해 update되는게 아니라 query encoder의 weight를 momentum하게 적용하여 update
10. Dictionary에 있는 queue를 방금 사용한 key representation을 넣어주고 오래된 representation은 내보냄

### 1.1. Contrastive Learning as Dictionary Look-up
<img src='https://user-images.githubusercontent.com/49643709/152975422-a0da9426-70c9-4a94-bd43-1e2b57d26bcb.png' width=80% height=80%>

Query의 representation 중 같은 이미지로부터 나온 key representation은 positive, dictionary에 있는 나머지 representation은 negative로 인식하여 InfoNCE loss로 학습

#### 1.2.1. Dictionary as a queue

Large dictionary를 이용하면 좋은 feature를 학습할 수 있다고 가정하는데, mini-batch의 경우 컴퓨팅 자원상 한계가 있음

Memory-bank 방식을 따라하되 약간의 변형을 주어 이용

Dictionary를 queue 자료구조형으로 만들어, encoded keys를 이용 후 dictionary에 저장

Queue 자료구조이기에, limit이 차면 예전 representations를 내보내고 새로운 encoded keys를 저장

#### 1.2.2. Momentum update
<img src='https://user-images.githubusercontent.com/49643709/152977242-cfe01001-ff2d-454e-ac56-98f47a173afc.png' width=80% height=80%>

Query encoder의 weight를 key encoder에 copy하면 key representation의 consistency가 감소하여 성능이 안좋음

이를 방지하기 위해 momentum update를 통해 key encoder update

Momentum을 크게 줘서 key encoder를 천천히 update하여 dictionary에 있는 representation과의 괴리를 줄일 수 있음
(momentum 0.999 performance > momentum 0.9 performance)

#### 1.2.3. Relation to previous mechanisms
![image](https://user-images.githubusercontent.com/49643709/152976092-4b4a3562-35a4-47b8-b5e3-e55d5c50c868.png)
- end-to-end
    - uses samples in the current mini-batch as the dictionary
    - keys are consistently encoded
    - dictionary size is coupled with the mini-batch size, limited by the GPU memory size
- memory bank
    - consists of the representations of all samples in the dataset
    - dictionary for each mini-batch is randomly sampled from the memory bank with no back-propagation
    - support large dictionary size
    - keys are less consistent

### 1.3. Pretext Task

Only use instance discrimination task

#### 1.3.1. Technical details

- ResNet with last fc layer with 128 dimension
- Output is normalized by its L2-norm (output = representation)
- Temperature is set as 0.07
- Data augmentation
    - random resize + random crop (224 * 224)
    - random color jittering
    - random horizontal flip
    - random grayscale conversion

#### 1.3.2. Shuffling BN
<img src='https://user-images.githubusercontent.com/49643709/153031378-01201c96-20b2-499a-85b9-22daa0237b9c.png' width=80% height=80%>

BN을 사용하면 low-loss solution을 쉽게 찾아 pretext task를 'cheat'하여 모델이 좋은 representation을 배울 수 없음

해당 이유로 BN이 samples간의 intra-batch communication을 통해 정보를 빼온다고 볼 수 있음

이를 해결하기 BN을 shuffling해서 해결해줌

Query는 따로 건드리지 않고 key mini-batch에서 sample order의 순서를 바꿔주어 representation을 구함

이렇게 하면 query와 key의 batch statistics가 달라지기에 해결됨

## 2. Experiments and Ablations

Experiment 세팅 값은 데이터셋마다 다르기에 자세한 값들은 논문 참고

### 2.1. Linear Classification Protocol

Encoder freeze해놓고 linear classifier만 붙여서 학습
(여기서 encoder는 ResNet 50의 global averaged feature임, 128 dimension representation이 아님)

다양한 실험에서 잘 작동되는 hyper-parameter 값을 찾았는데, 이는 기존의 supervised ImageNet과 다른 feature distribution을 가졌다는 것으로 볼 수 있음

#### 2.1.1. Ablation : contrastive loss mechanisms
<img src='https://user-images.githubusercontent.com/49643709/152981858-101c413a-0506-4ec6-85a6-155faf44bbe5.png' width=80% height=80%>

- Dictionary queue의 크기인 K가 커질수록 모든 mechanism이 performance 증가
- end-to-end mechanism은 dictionary size에 limit이 있고 large mini-batch optimization이 어려워 안좋음
- memory bank mechanism은 memory bank에 있는 representation이 consistent하지 않아 안좋음

#### 2.1.2. Ablation : momentum
<img src='https://user-images.githubusercontent.com/49643709/152982679-9bdbb473-004a-4054-be53-214175e5dfe9.png' width=80% height=80%>

Momentum이 커서 key encoder가 천천히 update되면 성능에 문제가 없음

Momentum 값이 너무 작아지면 성능이 안좋아지기도 하고 (m = 0.9), 극단적으로 가면 loss가 oscillation함 (m = 0)

이로부터 dictionary representation의 consistency가 중요하다는 것을 알 수 있음

#### 2.1.3. Comparison with previous results
<img src='https://user-images.githubusercontent.com/49643709/152983128-71ad339a-f0d2-4234-84af-03a8e3c98a0e.png' width=80% height=80%>

#### 2.2.1. Normalization

MoCo로 배운 distribution과 supervised ImageNet pre-training distribution은 다를 수 있음

Downstream task에서 이를 해결하기 위해 fine-tuning할 때 feature normalization을 진행

기존에 있던 BN을 학습시키며, 새로 초기화된 layer의 경우 BN을 사용하여 calibration을 돕게 해줌

#### 2.2.2. Schedules

Feature transferability를 확인하기 위해 schedule를 control해서 experiment 진행

최적이 아니지만, supervised pre-training과 똑같은 세팅으로 fine-tuning 진행

#### 2.3.1. Ablation : backbones
<img src='https://user-images.githubusercontent.com/49643709/153028125-a449986a-721d-44ca-8b18-59e409bdb67b.png' width=80% height=80%>

Transferring accuracy는 detector structure에 depend함

#### 2.3.2. Ablation : contrastive loss mechanisms
<img src='https://user-images.githubusercontent.com/49643709/153028598-dc8ff080-f221-48a2-a3a4-a8aeae0de8b1.png' width=80% height=80%>

#### 2.3.3. Comparison with previous results
![image](https://user-images.githubusercontent.com/49643709/153029418-d4afd224-aa94-454d-8a8e-b53a95ead933.png)

### 2.4. COCO Object Detection and Segmentation
![image](https://user-images.githubusercontent.com/49643709/153029584-066abf37-4c6d-42ae-a9ed-a943eee8df41.png)

### 2.5. More Downstream Tasks
<img src='https://user-images.githubusercontent.com/49643709/153029713-97b7fa89-f675-49a2-83c2-f32fc736585f.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/153030919-ca63b602-4b32-4782-a04a-dc34afac9a83.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/153030970-3824eda5-c3d9-4aaf-a6df-8ab3e99efdef.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/153031052-19d3d0a0-7e5b-4eb0-b059-bbf4630e7c39.png' width=80% height=80%>

![image](https://user-images.githubusercontent.com/49643709/153031145-633637af-1cb4-4843-af96-f0b5b9090cdf.png)

### 2.7. Summary

Segmentation task 몇몇에서는 성능이 다소 떨어지지만, 대부분의 downstream task에서 MoCo는 ImageNet supervised pre-training보다 더 좋은 성과를 보임

또한 IG-1B로 학습한 모델이 IN-1M보다 성능이 일관적으로 좋았기에 large-scale, uncurated dataset에 대해서도 잘 작동한다는 것을 확인할 수 있었음

**Comment by Jasonlee1995** [comment]
![image](https://user-images.githubusercontent.com/49643709/155917894-61a37502-a7bb-4f63-ab96-275935939d20.png)

Representation을 normalize한 뒤 matrix multiplication, softmax 적용 후 cross entropy loss 적용
